src/lib/comms/sol-netctl-impl-connman.c

```diff
@@ -0,0 +1,1133 @@
+        SOL_NULL_CHECK(e, -ENOMEM);
+    }
+
+    if (_ctx.properties_changed)
+        return 0;
+
+    r = snprintf(matchstr, sizeof(matchstr),
+        "type='signal',"
+        "interface='net.connman.Manager'");
+    SOL_INT_CHECK(r, < 0, r);
+
+    r = sd_bus_add_match(bus, &_ctx.properties_changed, matchstr,
+        _match_properties_changed, NULL);
+    SOL_INT_CHECK(r, < 0, r);
+
+    return 0;
+}
+
+static int
+dbus_service_del_monitor(
+    sol_connman_service_monitor_cb cb, const void *data)
+{
+    int r;
+
+    r = sol_monitors_find(&_ctx.service_ms, (sol_monitors_cb_t)cb, data);
+    SOL_INT_CHECK(r, < 0, r);
+
+    return sol_monitors_del(&_ctx.service_ms, r);
+}
+
+SOL_API int
+sol_connman_add_service_monitor(
+    sol_connman_service_monitor_cb cb, const void *data)
+{
+    int r;
+
+    r = sol_connman_init_lazy();
+    SOL_INT_CHECK_GOTO(r, < 0, fail_init);
+
+    r = dbus_service_add_monitor(cb, data);
+    SOL_INT_CHECK_GOTO(r, < 0, fail);
+
+    r = dbus_connection_get_service_properties();
+    SOL_INT_CHECK_GOTO(r, < 0, fail);
+
+    return 0;
+
+fail:
+    dbus_service_del_monitor(cb, data);
+
+fail_init:
+    sol_connman_shutdown_lazy();
+
+    return -EINVAL;
+}
+
+SOL_API int
+sol_connman_del_service_monitor(
+    sol_connman_service_monitor_cb cb, const void *data)
+{
+    SOL_NULL_CHECK(_ctx.connman, -EINVAL);
+
+    dbus_service_del_monitor(cb, data);
+    sol_connman_shutdown_lazy();
+
+    return 0;
+}
+
+static int
+dbus_manager_add_monitor(
+    sol_connman_manager_monitor_cb cb, const void *data)
+{
+    int r;
+    struct sol_monitors_entry *e;
+    char matchstr[256];
+    sd_bus *bus = sol_bus_client_get_bus(_ctx.connman);
+
+    SOL_NULL_CHECK(bus, -EINVAL);
+
+    r = sol_monitors_find(&_ctx.manager_ms, (sol_monitors_cb_t)cb, data);
+    if (r < 0) {
+        e = sol_monitors_append(&_ctx.manager_ms, (sol_monitors_cb_t)cb, data);
+        SOL_NULL_CHECK(e, -ENOMEM);
+    }
+
+    if (_ctx.properties_changed)
+        return 0;
+
+    r = snprintf(matchstr, sizeof(matchstr),
+        "type='signal',"
+        "interface='net.connman.Manager'");
+    SOL_INT_CHECK(r, < 0, r);
+
+    r = sd_bus_add_match(bus, &_ctx.properties_changed, matchstr,
+        _match_properties_changed, NULL);
+    SOL_INT_CHECK(r, < 0, r);
+
+    return 0;
+}
+
+static int
+dbus_manager_del_monitor(
+    sol_connman_manager_monitor_cb cb, const void *data)
+{
+    int r;
+
+    r = sol_monitors_find(&_ctx.manager_ms, (sol_monitors_cb_t)cb, data);
+    SOL_INT_CHECK(r, < 0, r);
+
+    return sol_monitors_del(&_ctx.manager_ms, r);
+}
+
+SOL_API int
+sol_connman_add_manager_monitor(
+    sol_connman_manager_monitor_cb cb, const void *data)
+{
+    int r;
+
+    r = sol_connman_init_lazy();
+    SOL_INT_CHECK_GOTO(r, < 0, fail_init);
+
+    r = dbus_manager_add_monitor(cb, data);
+    SOL_INT_CHECK_GOTO(r, < 0, fail);
+
+    r = dbus_connection_get_manager_properties();
```

Comment:
like above, only call get_manager_properties if it's the first one (ie: you just added the signal match)